### PR TITLE
tree: add FmtExport formatting for array types

### LINKED ISSUE #33429: importccl: EXPORT/IMPORT does not roundtrip arrays properly
```sql
> export into csv 'nodelocal:///foo.csv' from select array['a','b','c'];
  filename | rows | bytes
+----------+------+-------+
  n1.0.csv |    1 |    15
(1 row)


> import table zc(x text[]) csv data ('nodelocal:///foo.csv');
pq: "nodelocal:///foo.csv": row 1: parse "x" as STRING[]: array must be enclosed in { and }:
```

### PR DESCRIPTION
**Pull request by rohany**
Part of work for #44322.

Fixes #33429.

Release note (enterprise change): This PR fixes a bug where
export/import could not round trip arrays correctly.

### CODE REVIEW

**Comment by rohany** [comment]
Does nodelocal work in logictests? I want to add a test for #33429 but I'm not sure where to put it.

**Comment by knz** [comment]
I don't understand the approach here.

This introduces an extra format special case (yet _another_ text output format, really) and does not solve the problem for previously-EXPORTed data files.

Wouldn't it be simpler to 1) keep the output format as-is 2) change ParseDatumAs to recognize the `ARRAY[` prefix?

**Comment by rohany** [comment]
That sounds cleaner -- I didn't think about previously EXPORTed files.

**Comment by knz** [comment]
> Does nodelocal work in logictests? I want to add a test for #33429 but I'm not sure where to put it.

The test needs to be somewhere `pkg/ccl` in any case.

Also perhaps you can use a Go test instead of a logic test.

File: pkg/ccl/importccl/import_stmt_test.go [hunk 1]
```diff
@@ -1553,6 +1553,19 @@ func TestImportCSVStmt(t *testing.T) {
...
+	defer leaktest.AfterTest(t)()
+	ctx := context.Background()
+	baseDir := filepath.Join("testdata", "csv")
+	tc := testcluster.StartTestCluster(
+		t, 1, base.TestClusterArgs{ServerArgs: base.TestServerArgs{ExternalIODir: baseDir}})
+	defer tc.Stopper().Stop(ctx)
+	conn := tc.Conns[0]
+	sqlDB := sqlutils.MakeSQLRunner(conn)
+	sqlDB.Exec(t, `EXPORT INTO CSV 'nodelocal:///foo.csv' FROM SELECT ARRAY['a', 'b', 'c']`)
+	sqlDB.Exec(t, `IMPORT TABLE zc(x TEXT[]) CSV DATA ('nodelocal:///foo.csv/n1.0.csv')`)
```
Comment (knz) [inline pkg/ccl/importccl/import_stmt_test.go]: I'd run one additional check here to confirm that it contains the same data.

(Also perhaps do an array with other data types inside, such as array-of-json, and one `[]BYTES` array with some special characters inside the nested byte arrays.)

**Review by knz — COMMENTED** [review]
I think we need to support _both_ the `{}` and `array[]` syntaxes side-by-side, because `{}` is also used when importing pg dumps and I'm not sure we implement a separate parser for that.

**Comment by rohany** [comment]
RFAL -- working on this led me to uncover some other bugs (https://github.com/cockroachdb/cockroach/issues/44414), which I'll address in a follow up PR.

**Review by knz — COMMENTED** [review]
> Reviewed 1 of 4 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44389)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mjibson and @rohany)

___
*[pkg/sql/sem/tree/parse_array_test.go, line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44389#-LzfV12v21RWe-n17CYQ:-LzfV12v21RWe-n17CYR:b-qursl9) ([raw file](https://github.com/cockroachdb/cockroach/blob/73525c57df80866dcc2a1d2be43c5243db6501f1/pkg/sql/sem/tree/parse_array_test.go#L74)):*
> ```Go
> 		{`{\x07}`, types.String, Datums{NewDString(`x07`)}},
> 
> 		{`{日本語}`, types.String, Datums{NewDString(`日本語`)}},
> ```

You need to add a dozen of rows of unit tests to exercise all the combinations of the new syntax you're supporting: nested, with special characters,  with invalid matching parentheses, differnet data types, etc.
___
*[pkg/sql/sem/tree/parse_array_test.go, line 177 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44389#-LzfV4Fc514t2AV7AbpG:-LzfV4Fd8V_SStgFSRnZ:b-ub9uze) ([raw file](https://github.com/cockroachdb/cockroach/blob/73525c57df80866dcc2a1d2be43c5243db6501f1/pkg/sql/sem/tree/parse_array_test.go#L177)):*
> ```Go
> 		// It might be unnecessary to disallow this, but Postgres does.
> 		{`{he"lo}`, types.String, `could not parse "{he\"lo}" as type string[]: malformed array`},
> ```

1) I think your parser currently supports an array defined via `{1,2,3]` (mismatched bracket types) This should be rejected and a test for that added here.

2) I think it supports `[...]` without the `array` keyword. I don't think that's a good idea. There needs to be a unit test here to verify it's properly rejected.

3) we also want to reject invalid combinations of brackets, `array[{...}]` and `{array[...]}` should be rejected.

In fact, (2) and (3) together suggest to me that the top level of the parser should check for the `array` keyword as first thing, and then depending on whether it is found set a parameter which decides whether to either use `[`/`]` or `{`/`}` to accept array contents, and reject the other unknown characters.

**Comment by rohany** [comment]
Thanks for the pointers -- this is far from done (sorry about the message that it was RFAL, i tried to delete it, but it sent a notification anyway).

**Comment by rohany** [comment]
closing this in favor of another PR.

**Pull request closed without merging**